### DISCUSSION TITLE
How to evaluate the etcd value size

### DISCUSSION THREAD

**Discussion by lance5890** [etcd-io/etcd#16450] [Q&A] [answered]
1. As we can get a etcd value by its key like this :
```
etcdctl get  /kubernetes.io/secrets/xxx
```
2. But how can we evaluate the specified value size, as we can only see the value data, how can we get its size

**Comment by jmhbnz** [comment]
Hey @lance5890 - For retrieving the size of a key, how about writing the value to disk and just using `ls`?

For example:

```bash
 james  ~  D  e  bin   main                                                                                   16:16:02 
 ➜ ./etcdctl get --print-value-only /kubernetes.io/secrets/xxx > value.tmp


 james  ~  D  e  bin   main                                                                                   16:17:39 
 ➜ ls -lh
total 73M
drwx------ 3 james james 4.0K Aug 21 16:14 default.etcd
-rwxrwxr-x 1 james james  32M Aug 20 09:45 etcd
-rwxrwxr-x 1 james james  21M Aug 20 09:45 etcdctl
-rwxrwxr-x 1 james james  21M Aug 20 09:45 etcdutl
-rw-rw-r-- 1 james james   13 Aug 21 16:16 value.tmp

```

**Reply by lance5890** [reply]
yes, the value can be retrieved by this way.  I just wonder if or not the value size have been compacted when restored in etcd boltdb

**Reply by jmhbnz** [reply]
That's a fair question, I haven't looked into if bbolt is doing any kind of compression / storage optimsation for values stored. @ahrtr might have some knowledge on this part of it.

At the very least you could use the approach I mentioned above as a worst case scenario / highest possible estimate for an individual key value.

**Reply by ahrtr** [reply, marked as answer]
> I just wonder if or not the value size have been compacted when restored in etcd boltdb

bbolt doesn't compact the value. But the key & value stored in bbolt is different from the key & value what users see. The key stored in bbolt is actually a monotonically increasing revision, and the value stored in bbolt includes both the real key & value users see. Please read [kvstore_txn.go#L209-L232](https://github.com/etcd-io/etcd/blob/e2e17c75fe1006ea44b6ad793fa7b23f5e3546f4/server/storage/mvcc/kvstore_txn.go#L209-L232)


If you really want to get the value size programatically, you should use the etcd SDK (call the [Get](https://github.com/etcd-io/etcd/blob/e2e17c75fe1006ea44b6ad793fa7b23f5e3546f4/client/v3/kv.go#L49) method), and then get the value size using expression something like `len(resp.Kvs[0].Value)`.

**Reply by lance5890** [reply]
Thanks so much
